agent/stats.go

```diff
@@ -0,0 +1,90 @@
+package agent
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"cdr.dev/slog"
+)
+
+// ConnStats wraps a net.Conn with statistics.
+type ConnStats struct {
+	*ProtocolStats
+	net.Conn `json:"-"`
+}
+
+var _ net.Conn = new(ConnStats)
+
+func (c *ConnStats) Read(b []byte) (n int, err error) {
+	n, err = c.Conn.Read(b)
+	atomic.AddInt64(&c.RxBytes, int64(n))
+	return n, err
+}
+
+func (c *ConnStats) Write(b []byte) (n int, err error) {
+	n, err = c.Conn.Write(b)
+	atomic.AddInt64(&c.TxBytes, int64(n))
+	return n, err
+}
+
+type ProtocolStats struct {
+	NumConns int64 `json:"num_comms"`
+
+	// RxBytes must be read with atomic.
+	RxBytes int64 `json:"rx_bytes"`
+
+	// TxBytes must be read with atomic.
+	TxBytes int64 `json:"tx_bytes"`
+}
+
+var _ net.Conn = new(ConnStats)
+
+// Stats records the Agent's network connection statistics for use in
+// user-facing metrics and debugging.
+type Stats struct {
+	sync.RWMutex  `json:"-"`
+	ProtocolStats map[string]*ProtocolStats `json:"conn_stats,omitempty"`
```

Comment:
The JSON naming feels off here. Why call this `conn_stats`, have a `ConnStats` struct, but then this be `ProtocolStats`?